core/src/test/java/com/linecorp/armeria/client/HttpClientMaxConcurrentStreamTest.java

```diff
@@ -168,4 +177,225 @@ public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteA
+                                          .decorator((delegate, ctx, req) -> {
+                                              ctx.logBuilder().whenAvailable(RequestLogProperty.SESSION)
+                                                 .thenAccept(requestLog -> {
+                                                     connectionTimings.add(requestLog.connectionTimings());
+                                                 });
+                                              return delegate.execute(ctx, req);
+                                          })
+                                          .build();
+        final AtomicInteger opens = new AtomicInteger();
+        final AtomicInteger closes = new AtomicInteger();
+        connectionPoolListener = new ConnectionPoolListener() {
+            @Override
+            public void connectionOpen(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                       InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                opens.incrementAndGet();
+            }
+
+            @Override
+            public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                         InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                closes.incrementAndGet();
+            }
+        };
+        final List<CompletableFuture<AggregatedHttpResponse>> receivedResponses = new ArrayList<>();
+
+        final int numExpectedConnections = 6;
+        final int numRequests = MAX_CONCURRENT_STREAMS * numExpectedConnections;
+        final int numFailedRequests = MAX_CONCURRENT_STREAMS - 1;
+
+        clientFactory.eventLoopGroup().execute(() -> {
+            for (int j = 0; j < numRequests; j++) {
+                receivedResponses.add(client.get(PATH).aggregate());
+            }
+            // two more requests which fails due to server maxNumConnections
+            for (int j = 0; j < numFailedRequests; j++) {
+                receivedResponses.add(client.get(PATH).aggregate());
+            }
+        });
+
+        await().untilAsserted(() -> assertThat(responses).hasSize(numRequests));
+        assertThat(opens).hasValue(numExpectedConnections);
+        assertThat(connectionTimings.stream().filter(
+                timings -> timings.pendingAcquisitionDurationNanos() > 0))
+                .hasSize(numRequests + numFailedRequests - 1);
+
+        // Check exception thrown by responses
+        await().untilAsserted(() -> assertThat(receivedResponses.stream().filter(
+                CompletableFuture::isCompletedExceptionally)).hasSize(2));
+        receivedResponses.stream().filter(CompletableFuture::isCompletedExceptionally).forEach(
+                responseFuture -> assertThatThrownBy(responseFuture::get)
+                        .hasCauseInstanceOf(UnprocessedRequestException.class));
+    }
+
+    @Test
+    void exceededMaxStreamsForMultipleEventLoops() {
```

Comment:
Note to self: check if this has a potential to be flaky depending on how requests are assigned to event loops

Reply:
confirmed that `HeapBasedEventLoopState` guarantees this